ulog.cpp

```diff
@@ -0,0 +1,350 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2017  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "ulog.h"
+
+#include <assert.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "mainloop.h"
+#include "util.h"
+
+#define ULOG_HEADER_SIZE 16
+#define ULOG_MAGIC                               \
+    {                                            \
+        0x55, 0x4C, 0x6F, 0x67, 0x01, 0x12, 0x35 \
+    }
+
+#define NO_FIRST_MSG_OFFSET 255
+
+struct _packed_ ulog_msg_header {
+    uint16_t msg_size;
+    uint8_t msg_type;
+};
+
+ULog::ULog(const char *logs_dir)
+    : Endpoint{"ULog", false}
+{
+    assert(logs_dir);
+    _logs_dir = logs_dir;
+}
+
+ULog::~ULog()
+{
+}
+
+static bool _ulog_timeout_cb(void *data)
+{
+    ULog *ulog = static_cast<ULog *>(data);
+    return ulog->logging_start_timeout();
+}
+
+bool ULog::logging_start_timeout()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_START;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = _target_system_id;
+
+    mavlink_msg_command_long_encode(_system_id, MAV_COMP_ID_ALL, &msg, &cmd);
+    _send_msg(&msg, _target_system_id);
+
+    return true;
+}
+
+bool ULog::start()
+{
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    char path[1024];
+
+    if (_file != -1) {
+        log_warning("ULog already started");
+        return false;
+    }
+
+    snprintf(path, sizeof(path), "%s/%i-%02i-%02i_%02i-%02i-%02i.ulg", _logs_dir,
```

Comment:
_logs_dir + all the names can be more than 1024. Use asprintf here. It's not like we are starting and stop thousand times per second.